### DISCUSSION TITLE
SPUDSOFT_XLS emitter with bad formatting comparing prototype_excel_XLS emitter

### DISCUSSION THREAD

**Discussion by LeonardoIPM** [eclipse-birt/birt#1765] [Q&A] [answered]
I don't know if there are some options to configure my emitter better, but here is what is hapenning:

I was using prototype_excel_XLS emitter and it was almost fine, but it just don't render images, so we switched it for SPUDSOFT_XLS emitter.

But when rendering it with SPUDSOFT_XLS, the header isn't so well configured, it get disformatted a lot. 
1. The XLS doesn't create cells with width separately to header and footer, then the header and footer don't cover all the report width.
2. No cell have line break and the width of them are just weird, sometimes bigger, some times smaller.

(I'm using BIRT 4.15).

**Comment by speckyspooky** [comment]
Here we would need really an example. I have done lot of improment to the spudsoft-emitter.
For the cells you can activate the line break at report level and the spudsoft-emitter will use all specified width-informations based on column level.

So to get more specific topics we would need a kind of demo-report and some screens that we can folow your topics.

**Reply by LeonardoIPM** [reply]
Oks, I'll try to get a report example/screenshoots. 
Maybe I can try using 4.16 also to see how would be the results.
Thanks for helping!

**Reply by speckyspooky** [reply]
Every idea to improve our BIRT product is welocme.

Here you can find addition **"user properties"** which are currently supported at the spudsoft emitter,
to improve and change the behavior.

A specially I added the indent-behavior for spudsoft-emitter with BIRT 4.16 - this wasn't supported at the previous versions.

**User-Properies, README.MD:**
https://github.com/eclipse-birt/birt/blob/451b1bc26cac9c11d0a1e9efac6952dc0f686403/engine/uk.co.spudsoft.birt.emitters.excel/src/uk/co/spudsoft/birt/emitters/excel/README.md

**Comment by hvbtup** [comment]
It is explicitly not a goal of the spudsoft emitter to generate "pixel perfect" layout.
The goal is to create sheets that are usable for humans.
If you want to create a pixel-perfect layout, use the PDF emitter.

Hint: You can merge cells, if you need different column widths for different rows.

**Comment by LeonardoIPM** [comment]
It was not a really big issue. We will continue using the old XLS emitter to don't change our reports behavior. Thanks for the answers!

**Reply by speckyspooky** [reply]
I'm interested to here what the requirement is.
May be it make sense to have it like a standard option of the BIRT emitter.

**Reply by LeonardoIPM** [reply]
Oks, I tried to simulate the spudsoft_emitter and the xls_emitter just to show the behavior.

xls_emitter doesn't print the image but have a good formatation
![OldEmitter](https://github.com/user-attachments/assets/1178402f-dcf6-4665-af97-fe2cb5d8f196)

spudsoft_emitter print the image but have a bad formatation, the cells aren't merged and the text don't have a line break (except the header).
Structured Header: ON
![Spudsoft](https://github.com/user-attachments/assets/3c12c740-6565-47ed-8745-841c0f090964)

**Reply by speckyspooky** [reply, marked as answer]
You can get the line breaks at the excel (Spudsoft) as well.
The topic is that the "whitespace"-handling is responsible of it.

So you can set in generally the "preformatted"-option (see screen).
But be aware the usage of this flag as directly an impact of all outputs.

![grafik](https://github.com/user-attachments/assets/810f9679-7b88-4b5c-aeef-6496c3a204f0)

Therefore I use instead the standard option a small script at the "onRender"-Method of the cell:

```
	if ( reportContext.getOutputFormat().indexOf("xls") >= 0 ){
		this.getStyle().whiteSpace = "pre";
	}
```

**Reply by speckyspooky** [reply]
The images is a bigger challenge because the merged rows/columns will scale the image automatically.
With BIRT 4.17 I optimzed the handling of images at a standard cell.

The handling of merged cells is more complex and not so easy to change.
